### PR TITLE
fixes #9270 JMS request reply capability

### LINKED ISSUE #9270: support for JMS request/reply
**Description:**
There is currently no way to implement JMS Request/Reply through ballerina. To make this possible I would recommend implementing a QueueRequestor (maybe TopicRequestor too), to handle this kind of task.

Sadly the JMS QueueRequestor class is not quite usable, since it does not implement a timeout for the call. To support timeouts, the implementation needs to contruct and handle the JMS resources manually.
sample java code:
```
QueueSession session = connection.createSession();
Queue reqQueue = session.createQueue(queueName);
QueueSender sender = session.createSender(reqQueue);
Queue replyQ = session.createTemporaryQueue();
MessageConsumer replyConsumer = session.createConsumer(replyQ);
TextMessage message = session.createTextMessage();
message.setJMSReplyTo(replyQ);
message.setText(msg);
sender.send(message);
Message replyMsg = replyConsumer.receive(RESPONSETIMEOUT);
```

I would see something like the following syntax to simplify the procedure in ballerina:
```
endpoint jms:QueueRequestor queueRequestor {
    session:jmsSession,
    queueName:"MyQueue",
    requestTimeout: 5000
};
function main(string... args) {
    match (jmsSession.createTextMessage("Hello from Ballerina")) {
        error e => {
            log:printError("Error occurred while creating message", err=e);
        }
        jms:Message msg => {
            var response = queueRequestor->request(msg);
            match (response) {
                jms:Message msg => {
                    log:printInfo("success");
                } 
                error err => {
                    log:printInfo("Error occurred while sending message");
                } 
            }
        }
    }
}
```

**Affected Versions:**
0.975.0

**Suggested Labels (optional):**
component/stdlib/jms, Type/New Feature

**Suggested Assignees (optional):**
I would start working on this.

### PR DESCRIPTION
**Pull request by JensWalter**
## Purpose
Provide request/reply capabilities to the JMS endpoint implementation as discussed in ticket #9270 .

There is still something missing I couldn't get to work.
The timeout parameter is given to the receiveFrom function as a parameter with a default value. I couldn't figure out how to read this value. This needs to be fixed before it is merged into master.

file: org.ballerinalang.net.jms.nativeimpl.endpoint.common.ReceiveFromActionHandler
line: 63

## Samples
I used the following samples to work with the implementation:

JMS Queue Receiver:
```
import ballerina/io;
import ballerina/jms;

jms:Connection jmsConnection = new ({   
    initialContextFactory:"com.tibco.tibjms.naming.TibjmsInitialContextFactory",
    providerUrl:"tcp://localhost:7222",
    connectionFactoryName: "QueueConnectionFactory",
    username:"admin",
    password:"admin",
    properties:{
        "java.naming.security.principal":"admin",
        "java.naming.security.credentials":"admin"
    }
});

jms:Session jmsSession = new (jmsConnection, {
    acknowledgementMode: "AUTO_ACKNOWLEDGE"
});

endpoint jms:QueueReceiver consumer {
    session: jmsSession,
    queueName: "test.queue.a"
};

endpoint jms:QueueSender jmsProducer {
    session: jmsSession
};

service<jms:Consumer> jmsListener bind consumer {
    onMessage (endpoint ep_consumer, jms:Message message) {
        io:println("receiving msg");
        io:println("building response1");
        var replyDest = message.getReplyTo(); 
        io:println("building response2");
        io:println(replyDest);
        io:println("building response3");
        match(replyDest){
            jms:Destination replyTo =>{ 
                var x = jmsSession.createTextMessage("Response text");
                match(x){
                    jms:Message msg => {
                    io:println("reply to "+replyTo.destinationName);
                        io:println("sending...");
                        var y = jmsProducer->sendTo(replyTo, msg);
                        io:println("done...");
                        io:println(y);
                    }
                    error e => io:println(e);
                }
            }
            error e =>{ io:println(e);}
            () => io:println("nothing found");
        }
      
    }
}
```

JMS Queue Requestor:
```
import ballerina/io;
import ballerina/jms;

jms:Connection jmsConnection = new ({   
    initialContextFactory:"com.tibco.tibjms.naming.TibjmsInitialContextFactory",
    providerUrl:"tcp://localhost:7222",
    connectionFactoryName: "QueueConnectionFactory",
    username:"admin",
    password:"admin",
    properties:{
        "java.naming.security.principal":"admin",
        "java.naming.security.credentials":"admin"
    }
});

jms:Session jmsSession = new (jmsConnection, {
    acknowledgementMode: "AUTO_ACKNOWLEDGE"
});

endpoint jms:QueueReceiver consumer {
    session: jmsSession
};

endpoint jms:QueueSender jmsProducer {
    session: jmsSession
};

function main(string... args){
    jms:Destination replyDest = check jmsSession.createTemporaryQueue();
    jms:Destination requestDest = check jmsSession.createQueue("test.queue.a");
    io:println("created destination");
    var x = jmsSession.createTextMessage("test request");
    io:println("created message");
    match(x) {
        jms:Message m => {
            _ = m.setReplyTo(replyDest);
            io:println("set reply");
            _ = jmsProducer->sendTo(requestDest,m);
            io:println("wait for response");
    
            var replyMsg = consumer->receiveFrom(replyDest,timeoutInMilliSeconds=5000);
            io:println(replyMsg);
            io:println("done.");
        }
        error err => {}
    }

}
```

JMS Topic Receiver:
```
import ballerina/io;
import ballerina/jms;

jms:Connection jmsConnection = new ({   
    initialContextFactory:"com.tibco.tibjms.naming.TibjmsInitialContextFactory",
    providerUrl:"tcp://localhost:7222",
    connectionFactoryName: "TopicConnectionFactory",
    username:"admin",
    password:"admin",
    properties:{
        "java.naming.security.principal":"admin",
        "java.naming.security.credentials":"admin"
    }
});

jms:Session jmsSession = new (jmsConnection, {
    acknowledgementMode: "AUTO_ACKNOWLEDGE"
});

endpoint jms:TopicSubscriber consumer {
    session: jmsSession,
    topicPattern: "test.topic.a"
};

endpoint jms:TopicPublisher jmsProducer {
    session: jmsSession
};

service<jms:Consumer> jmsListener bind consumer {
    onMessage (endpoint ep_consumer, jms:Message message) {
        io:println("receiving msg");
        io:println("building response1");
        var replyDest = message.getReplyTo(); 
        io:println("building response2");
        io:println(replyDest);
        io:println("building response3");
        match(replyDest){
            jms:Destination replyTo =>{ 
                var x = jmsSession.createTextMessage("Response text");
                match(x){
                    jms:Message msg => {
                    io:println("reply to "+replyTo.destinationName);
                        io:println("sending...");
                        var y = jmsProducer->sendTo(replyTo, msg);
                        io:println("done...");
                        io:println(y);
                    }
                    error e => io:println(e);
                }
            }
            error e =>{ io:println(e);}
            () => io:println("nothing found");
        }
      
    }
}

```
JMS Topic Requestor:
```
import ballerina/io;
import ballerina/jms;

jms:Connection jmsConnection = new ({   
    initialContextFactory:"com.tibco.tibjms.naming.TibjmsInitialContextFactory",
    providerUrl:"tcp://localhost:7222",
    connectionFactoryName: "TopicConnectionFactory",
    username:"admin",
    password:"admin",
    properties:{
        "java.naming.security.principal":"admin",
        "java.naming.security.credentials":"admin"
    }
});

jms:Session jmsSession = new (jmsConnection, {
    acknowledgementMode: "AUTO_ACKNOWLEDGE"
});

endpoint jms:TopicSubscriber consumer {
    session: jmsSession,
    topicPattern: "test.topic.a"
};

endpoint jms:TopicPublisher jmsProducer {
    session: jmsSession
};

function main(string... args){
    jms:Destination replyDest = check jmsSession.createTemporaryTopic();
    jms:Destination requestDest = check jmsSession.createTopic("test.topic.a");
    var x = jmsSession.createTextMessage("test request");
    match (x){
        jms:Message m => {
            _ = m.setReplyTo(replyDest);
            _ = jmsProducer -> sendTo(requestDest,m);
            var replyMsg = consumer->receiveFrom(replyDest,timeoutInMilliSeconds=5000);
            match (replyMsg) {
                jms:Message m2 => {
                    io:println("got response message");
                    var t = m2.getTextMessageContent();
                    match(t){
                        string str => io:println(str);
                        error err =>{}
                       
                    }
                }
                error err => {}
                () => {}
            }
            io:println("done.");
        }
        error e => {}
    }
   

}
```

### CODE REVIEW

File: stdlib/jms/src/main/java/org/ballerinalang/net/jms/nativeimpl/endpoint/common/ReceiveFromActionHandler.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                consumer.close();
+            } else {
+                context.setReturnValues();
+            }
+        } catch (JMSException e) {
+            BallerinaAdapter.returnError("Message receiving failed.", context, e);
+        }
+    }
+}
+
```
Comment (lahirus) [inline stdlib/jms/src/main/java/org/ballerinalang/net/jms/nativeimpl/endpoint/common/ReceiveFromActionHandler.java]: Some unnecessary new lines.

File: stdlib/jms/src/main/ballerina/jms/destination.bal [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+// under the License.
+
+documentation { Destination object
+    F{{destinationName}} name of the destination
+    F{{destinationType}} type of the destination, either queue or topic.
+}
+public type Destination object {
+    public string destinationName;
+    public string destinationType;
+};
```
Comment (Shan1024) [inline stdlib/jms/src/main/ballerina/jms/destination.bal]: No newline at the end of the file.

File: stdlib/jms/src/main/java/org/ballerinalang/net/jms/nativeimpl/endpoint/common/CloseConsumerHandler.java [hunk 3]
```diff
@@ -38,14 +38,18 @@ private CloseConsumerHandler() {
...
-                                                                    context);
-        try {
-            consumer.close();
-        } catch (JMSException e) {
-            BallerinaAdapter.throwBallerinaException("Error closing message consumer.", context, e);
+        if (connectorBObject.getNativeData(Constants.JMS_CONSUMER_OBJECT) != null) {
+            MessageConsumer consumer = BallerinaAdapter.getNativeObject(connectorBObject,
+                                                                        Constants.JMS_CONSUMER_OBJECT,
+                                                                        MessageConsumer.class,
+                                                                        context);
```
Comment (Shan1024) [inline stdlib/jms/src/main/java/org/ballerinalang/net/jms/nativeimpl/endpoint/common/CloseConsumerHandler.java]: We can move the content of this line to the previous line.

**Comment by wggihan** [comment]
@JensWalter We have fixed the conflicts and formatting issues with this [9860](https://github.com/ballerina-platform/ballerina-lang/pull/9860). We will merge that PR once we review it for the functional compatibility. We will close this PR. Thanks for the contribution.

**Pull request closed without merging**